bundles/org.eclipse.emfcloud.modelserver.client/src/org/eclipse/emfcloud/modelserver/client/AbstractModelServerClient.java

```diff
@@ -0,0 +1,807 @@
+   }
+
+   public boolean send(final String modelUri, final String message) {
+      final WebSocket webSocket = openSockets.get(modelUri);
+      if (webSocket != null) {
+         return webSocket.send(message);
+      }
+      return false;
+   }
+
+   public boolean unsubscribe(final String modelUri) {
+      final WebSocket webSocket = openSockets.get(modelUri);
+      if (webSocket != null) {
+         return webSocket.close(1000, "Websocket closed by client.");
+      }
+      return false;
+   }
+
+   protected String makeWsUrl(final String path) {
+      return path.replaceFirst("http:", "ws:");
+   }
+
+   protected String makeUrl(final String path) {
+      return baseUrl + path;
+   }
+
+   protected HttpUrl.Builder createHttpUrlBuilder(final String path) {
+      return Objects.requireNonNull(HttpUrl.parse(path)).newBuilder();
+   }
+
+   protected CompletableFuture<Response<String>> makeCall(final Request request) {
+      CompletableFuture<Response<String>> future = new CompletableFuture<>();
+      this.client.newCall(request).enqueue(new Callback() {
+         @Override
+         public void onFailure(@NotNull final Call call, @NotNull final IOException e) {
+            future.completeExceptionally(e);
+         }
+
+         @Override
+         public void onResponse(@NotNull final Call call, @NotNull final okhttp3.Response response) {
+            future.complete(new Response<>(response));
+         }
+      });
+
+      return future;
+   }
+
+   protected Response<Optional<String>> parseField(final Response<String> response, final String field) {
+      return response.mapBody(body -> parseJsonField(body, field));
+   }
+
+   protected Optional<String> parseJsonField(final String jsonAsString, final String field) {
+      try {
+         final JsonNode node = Json.parse(jsonAsString);
+         return getJsonField(node, field);
+      } catch (IOException e) {
+         LOG.error("Could not parse JSON", e);
+         LOG.debug("Invalid JSON: " + jsonAsString);
+         return Optional.empty();
+      }
+   }
+
+   protected Optional<String> getJsonField(final JsonNode node, final String field) {
+      JsonNode data = node.get(field);
+      if (data == null) {
+         return Optional.empty();
+      }
+      if (data.isTextual()) {
+         return Optional.of(data.textValue());
+      }
+      return Optional.of(data.toString());
+   }
+
+   protected <A> Response<A> getBodyOrThrow(final Response<Optional<A>> response) {
+      return response.mapBody(this::require);
+   }
+
+   protected <A> A require(final Optional<A> value) {
+      return value.orElseThrow(() -> new RuntimeException("Could not parse 'data' field"));
+   }
+
+   public String encode(final EObject eObject) {
+      return encode(eObject, getDefaultFormat());
+   }
+
+   public String encode(final EObject eObject, final String format) {
+      try {
+         if (format.equals(ModelServerPathParametersV1.FORMAT_XMI)) {
```

Comment:
I'm not sure it really makes a difference, since we only expect XMI or Json (We don't even check registered Codecs here). But we should probably add support for decoding Json-V2 objects (Currently, we only use Json-V2 for encoding Models)